google/cloud/bigtable/mutations_batcher.py

```diff
@@ -46,61 +174,221 @@ class MutationsBatcher:
         """
-        raise NotImplementedError
+        Flush all staged mutations
 
-    async def append_conditional(
+        Args:
+          - raise_exceptions: if True, will raise any unreported exceptions from this or previous flushes.
+              If False, exceptions will be stored in self.exceptions and raised on a future flush
+              or when the batcher is closed.
+          - timeout: maximum time to wait for flush to complete, in seconds.
+              If exceeded, flush will continue in the background and exceptions
+              will be surfaced on the next flush
+        Raises:
+          - MutationsExceptionGroup if raise_exceptions is True and any mutations fail
+          - asyncio.TimeoutError if timeout is reached before flush task completes.
+        """
+        # add recent staged mutations to flush task, and wait for flush to complete
+        flush_job : Awaitable[None] = self._schedule_flush()
+        if timeout is not None:
+            # wait `timeout seconds for flush to complete
+            # if timeout is exceeded, flush task will still be running in the background
+            flush_job = asyncio.wait_for(asyncio.shield(flush_job), timeout=timeout)
+        await flush_job
+        # raise any unreported exceptions from this or previous flushes
+        if raise_exceptions:
+            self._raise_exceptions()
+
+    def _schedule_flush(self) -> asyncio.Task[None]:
+        """Update the flush task to include the latest staged mutations"""
+        if self._staged_mutations:
+            entries, self._staged_mutations = self._staged_mutations, []
+            self._staged_count, self._staged_bytes = 0, 0
+            self._prev_flush = asyncio.create_task(
+                self._flush_internal(entries, self._prev_flush)
+            )
+        return self._prev_flush
+
+    async def _flush_internal(
         self,
-        predicate_filter: RowFilter,
-        row_key: str | bytes,
-        if_true_mutations: Mutation | list[Mutation] | None = None,
-        if_false_mutations: Mutation | list[Mutation] | None = None,
+        new_entries: list[RowMutationEntry],
+        prev_flush: asyncio.Task[None],
     ):
         """
-        Apply a different set of mutations based on the outcome of predicate_filter
+        Flushes a set of mutations to the server, and updates internal state
 
-        Calls check_and_mutate_row internally on flush
+        Args:
+          - new_entries: list of mutations to flush
+          - prev_flush: the previous flush task, which will be awaited before
+              a new flush is initiated
         """
-        raise NotImplementedError
+        # wait for previous flush to complete
+        await prev_flush
+        # flush new entries
+        async for batch in self._flow_control.add_to_flow(new_entries):
+            batch_errors = await self._execute_mutate_rows(batch)
```

Comment:
Doesnt this prevent concurrency? The idea of a batcher is that it should break up a stream of mutations into X concurrent batches of size Y and when X * Y limit is reached, it applies back pressure to the caller or queues the mutations.